lib/Conversion/TorchToLinalg/DataMovement.cpp

```diff
@@ -334,139 +318,279 @@ class ConvertAtenViewOp : public OpConversionPattern<AtenViewOp> {
+                op,
+                "found ambiguous association of dynamic sizes");
+          }
+          outputAssociations.back().push_back(outputDim);
+          if (failed(collapseToSingleDim(
+                  op, rewriter, outputDim, nextUnchangedOutput, inputDim,
+                  nextUnchangedInput, outputShape, inputShapeVec,
+                  inputAssociations.back())))
+            return failure();
+          continue;
         }
-        // To meet both requirements from tensor.expand_shape verification code.
-        collapsedShape[collapsedDim] = collapsedDimSize;
-        expandedDim = expandedDimNext;
-        continue;
-      }
 
-      // collpasedDims are expanded to [expandedDim, expandedDimNext)
-      if (expandedDimNext - expandedDim < (int64_t)collapsedDims.size())
-        op.emitError("unimplemented: mixed of expanding and collapsing "
-                     "operations for view");
-      for (auto collapsedDim : collapsedDims) {
-        if (collapsedShape[collapsedDim] == kUnknownSize) {
-          if (expandedDim >= expandedDimNext) {
+        // inputDim is next to the boundary
+        if (inputDim == nextUnchangedInput - 1) {
+          if (hasDynamic && inputShape[inputDim] == kUnknownSize) {
             return rewriter.notifyMatchFailure(
                 op,
-                "desired size is not compatible with the input tensor size");
+                "found ambiguous association of dynamic sizes");
           }
-          checkDimEqualHelper(rewriter, loc, collapsedShapeInt[collapsedDim],
-                              expandedShapeInt[expandedDim]);
-          // To meet the second requirement from tensor.expand_shape
-          // verification code.
-          expandedShape[expandedDim] = kUnknownSize;
-          reassociation[collapsedDim].push_back(expandedDim++);
-        } else {
-          int64_t remainingSizeToExpand = collapsedShape[collapsedDim];
-          // A do-while loop is used here to handle the cases where the
-          // collapsed shape tensor has a dimension of size 1.
-          do {
-            int64_t expandedDimSize = expandedShape[expandedDim];
-            if (expandedDim >= expandedDimNext ||
-                expandedShape[expandedDim] == kUnknownSize ||
-                remainingSizeToExpand % expandedDimSize != 0) {
+          inputAssociations.back().push_back(inputDim);
+          if (failed(collapseToSingleDim(
+                  op, rewriter, inputDim, nextUnchangedInput, outputDim,
+                  nextUnchangedOutput, inputShapeVec, outputShape,
+                  outputAssociations.back())))
+            return failure();
+          continue;
+        }
+
+        // Helper to find the minimum set of dims to collapse with the
+        // same number of elements as that of collapseDim. This function assumes
+        // the size of the collapsed dim is never dynamic.
+        auto minimallyCollapseDim = [](AtenViewOp op,
```

Comment:
nit: can you make this helper into a `static` function? The lowering for this op keeps getting longer and longer, which makes it harder to read